contracts/contribute/Contributors.sol

```diff
@@ -324,83 +324,94 @@ contract Contributors is ERC721TokenReceiver {
+
+        // Set agent params
+        agentParams = new IService.AgentParams[](NUM_AGENT_INSTANCES);
+        agentParams[0] = IService.AgentParams(uint32(NUM_AGENT_INSTANCES), uint96(minStakingDeposit));
+    }
+
+    /// @dev Activates service agent registration and registers agent instance.
+    /// @param serviceId Minted service Id.
+    /// @param minStakingDeposit Minimum service security deposit.
+    /// @param agentIds Service agent Ids.
+    function _activateServiceRegisterAgentInstance(
+        uint256 serviceId,
+        uint256 minStakingDeposit,
+        uint32[] memory agentIds
+    ) internal {
         // Set agent instances as [msg.sender]
         address[] memory instances = new address[](NUM_AGENT_INSTANCES);
         instances[0] = msg.sender;
 
-        // Create a service owned by this contract
-        serviceId = IService(serviceManager).create(address(this), token, configHash, agentIds,
-            agentParams, uint32(THRESHOLD));
+        // Calculate the total bond required for the service deployment
+        uint256 totalBond = (1 + NUM_AGENT_INSTANCES) * minStakingDeposit;
+
+        // Transfer the total bond amount from the contributor
+        IToken(olas).transferFrom(msg.sender, address(this), totalBond);
+        // Approve token for the serviceRegistryTokenUtility contract
+        IToken(olas).approve(serviceRegistryTokenUtility, totalBond);
 
         // Activate registration (1 wei as a deposit wrapper)
         IService(serviceManager).activateRegistration{value: 1}(serviceId);
 
         // Register msg.sender as an agent instance (numAgentInstances wei as a bond wrapper)
         IService(serviceManager).registerAgents{value: NUM_AGENT_INSTANCES}(serviceId, instances, agentIds);
-
-        // Prepare Safe multisig data
-        uint256 localNonce = _nonce;
-        uint256 randomNonce = uint256(keccak256(abi.encodePacked(block.timestamp, msg.sender, localNonce)));
-        bytes memory data = abi.encodePacked(address(0), fallbackHandler, address(0), address(0), uint256(0),
-            randomNonce, "0x");
-        // Deploy the service
-        multisig = IService(serviceManager).deploy(serviceId, safeMultisig, data);
-
-        // Update the nonce
-        _nonce = localNonce + 1;
     }
 
+    /// @dev Re-deploys the service.
+    /// @param serviceId Service Id.
+    /// @param stakingInstance Staking instance address.
+    /// @param multisig Corresponding service multisig.
+    /// @param updateService True if service update is required.
     function _reDeploy(uint256 serviceId, address stakingInstance, address multisig, bool updateService) internal {
-        // Get deposit service info for staking
-        uint256 minStakingDeposit = IStaking(stakingInstance).minStakingDeposit();
-
-        // Set agent Ids
-        uint32[] memory agentIds = new uint32[](NUM_AGENT_INSTANCES);
-        agentIds[0] = uint32(agentId);
+        // Get and check service params
```

Comment:
The workflow is then consists of 4 parts:
- Get service params;
- Create a new service or update one, if needed;
- Activate and register staking instance;
- Deploy a new service or re-deploy the existing one.